framework/src/problems/FEProblemBase.C

```diff
@@ -3954,6 +3961,64 @@ FEProblemBase::execute(const ExecFlagType & exec_type)
 
   // Return the current flag to None
   setCurrentExecuteOnFlag(EXEC_NONE);
+
+  if (_uo_aux_state_check && !_checking_uo_aux_state)
+  {
+    // we will only check aux variables and postprocessors
+    // checking more reporter data can be added in the future if needed
+    std::unique_ptr<NumericVector<Number>> x = _aux->currentSolution()->clone();
+    DenseVector<Real> pp_values = getReporterData().getAllPostprocessorValues();
+
+    // call THIS execute one more time for checking the possible states
+    _checking_uo_aux_state = true;
+    FEProblemBase::execute(exec_type);
+    _checking_uo_aux_state = false;
+
+    *x -= *_aux->currentSolution();
+    if (x->l2_norm() > 1e-8)
+    {
+      const auto & sys = _aux->system();
+      unsigned int n_vars = sys.n_vars();
+      std::map<Real, std::string, std::greater<Real>> ordered_map;
+      for (unsigned int i = 0; i < n_vars; i++)
+        ordered_map[sys.calculate_norm(*x, i, DISCRETE_L2)] = sys.variable_name(i);
+
+      std::ostringstream oss;
+      for (const auto & pair : ordered_map)
+        oss << "  {" << pair.second << ", " << pair.first << "},\n";
+
+      mooseError("Aux kernels, user objects appear to have states for aux variables on ",
+                 exec_type,
+                 ".\nVariable error norms in descending order:\n",
+                 oss.str());
+    }
+
+    pp_values -= getReporterData().getAllPostprocessorValues();
```

Comment:
Should probably check to size match before doing this operation.

Reply:
I will add an assert. Supposedly after the second call of `execute()` at the same execute, the size should not change.

Reply:
I changed my mind, I will just add a check because this is not on performance critical code path and can make the code more protective.